### PR TITLE
Save api requests to MongoDB

### PR DESCRIPTION
**Pull request by tabladrum**

### CODE REVIEW

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+            String line = null;
+            StringBuilder inputBuffer = new StringBuilder();
+            do {
+                line = reader.readLine();
+                if (null != line) {
+                    inputBuffer.append(line.trim());
+                }
+            } while (line != null);
+            reader.close();
+            return inputBuffer.toString().trim();
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: at this point there are several copies of the body in memory.. we should rethink & optimize this..

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+            BufferedReader reader = new BufferedReader(new InputStreamReader(this.getInputStream()));
+            String line = null;
+            StringBuilder inputBuffer = new StringBuilder();
+            do {
+                line = reader.readLine();
+                if (null != line) {
+                    inputBuffer.append(line.trim());
+                }
+            } while (line != null);
+            reader.close();
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: use try-with-resource

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+                .append(httpServletRequest.getRemoteAddr())
+                .append("] [STATUS:")
+                .append(bufferedResponse.getStatus())
+                .append("]");
+        LOGGER.info(logMessage.toString());
+
+        if (httpServletRequest.getMethod().equals(HttpMethod.PUT.toString()) ||
+                (httpServletRequest.getMethod().equals(HttpMethod.POST.toString())) ||
+                (httpServletRequest.getMethod().equals(HttpMethod.DELETE.toString()))) {
+            requestLogRepository.save(requestLog);
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: try / catch cause if u cant save u dont want to impact the request/response

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+                .append("] [PARAMETERS:")
+                .append(requestMap)
+                .append("] [BODY:")
+                .append(bufferedRequest.getRequestBody())
+                .append("] [REMOTE:")
+                .append(httpServletRequest.getRemoteAddr())
+                .append("] [STATUS:")
+                .append(bufferedResponse.getStatus())
+                .append("]");
+        LOGGER.info(logMessage.toString());
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: don't toString unless u absolutely need to.. i dont think u want to dump the body at info, maybe only if mongo save fails..

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java (hunk 2 above)
Comment (tabladrum) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: Good idea!

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+
+
+        public BufferedRequestWrapper(HttpServletRequest req) throws IOException {
+            super(req);
+            // Read InputStream and store its content in a buffer.
+            InputStream is = req.getInputStream();
+            this.baos = new ByteArrayOutputStream();
+            byte buf[] = new byte[1024];
+            int letti;
+            while ((letti = is.read(buf)) > 0) {
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: should probably cap this

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+
+        HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+        HttpServletResponse httpServletResponse = (HttpServletResponse) response;
+
+        Map<String, String> requestMap = this.getTypesafeRequestMap(httpServletRequest);
+        BufferedRequestWrapper bufferedRequest = new BufferedRequestWrapper(httpServletRequest);
+        BufferedResponseWrapper bufferedResponse = new BufferedResponseWrapper(httpServletResponse);
+
+
+        RequestLog requestLog = new RequestLog();
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: so if the request isnt PUT, POST, DELETE - i wouldnt even do this stuff, just immediately pass it on to do the next filter, else do this other stuff

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+        if (settings.isLogRequest()) {
+            final StringBuilder logMessage = new StringBuilder("REST Request - ")
+                    .append("[")
+                    .append(httpServletRequest.getMethod())
+                    .append("] [PATH:")
+                    .append(httpServletRequest.getPathInfo())
+                    .append("] [PARAMETERS:")
+                    .append(requestMap)
+                    .append("] [BODY:")
+                    .append(bufferedRequest.getRequestBody())
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: if the request content-type isnt what u expect, json? then probably shouldnt do this.. if for some reason it's octet-stream, dont bother..

Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: should also capture content-type

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+            String line = null;
+            StringBuilder inputBuffer = new StringBuilder();
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(this.getInputStream()))) {
+                do {
+                    line = reader.readLine();
+                    if (null != line) {
+                        inputBuffer.append(line.trim());
+                    }
+                } while (line != null);
+                reader.close();
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: close is done with the try-with-resource

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+        }
+
+        String getRequestBody() throws IOException {
+            String line = null;
+            StringBuilder inputBuffer = new StringBuilder();
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(this.getInputStream()))) {
+                do {
+                    line = reader.readLine();
+                    if (null != line) {
+                        inputBuffer.append(line.trim());
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: why trim it? dont u want to see exactly what they posted?

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+                    }
+                } while (line != null);
+                reader.close();
+            }
+            return inputBuffer.toString().trim();
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: why trim it? dont u want to see exactly what they posted?

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+                    .append(requestMap)
+                    .append("] [BODY:")
+                    .append(bufferedRequest.getRequestBody())
+                    .append("] [REMOTE:")
+                    .append(httpServletRequest.getRemoteAddr())
+                    .append("] [STATUS:")
+                    .append(bufferedResponse.getStatus())
+                    .append("]");
+
+            if (httpServletRequest.getMethod().equals(HttpMethod.PUT.toString()) ||
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: this check is redundant now

Comment (tabladrum) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: duh!!!

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+            requestLog.setParameter(requestMap.toString());
+            requestLog.setRequestSize(httpServletRequest.getContentLengthLong());
+            requestLog.setRequestContentType(httpServletRequest.getContentType());
+            if (httpServletRequest.getContentType().startsWith(APPLICATION_JSON_VALUE)) {
+                requestLog.setRequestBody(JSON.parse(bufferedRequest.getRequestBody()));
+            }
+
+            chain.doFilter(bufferedRequest, bufferedResponse);
+            requestLog.setResponseContentType(httpServletResponse.getContentType());
+            if (httpServletResponse.getContentType().startsWith(APPLICATION_JSON_VALUE)) {
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: to be safe i'd use something like: ```new MimeType(..).isCompatibleWith(APPLICATION_JSON_VALUE)```

File: api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+                requestLog.setResponseBody(JSON.parse(bufferedResponse.getContent()));
+            }
+            requestLog.setResponseSize(bufferedResponse.getContent().length());
+
+            requestLog.setResponseCode(bufferedResponse.getStatus());
+            requestLog.setTimestamp(System.currentTimeMillis());
+            final StringBuilder logMessage = new StringBuilder("REST Request - ")
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/config/LoggingFilter.java]: if this only used in the else, should just move it there

File: core/src/main/java/com/capitalone/dashboard/model/RequestLog.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
+package com.capitalone.dashboard.model;
+
+import org.springframework.data.mongodb.core.mapping.Document;
+
+
+@Document(collection = "requests")
+public class RequestLog extends BaseModel {
```
Comment (jdamick) [inline core/src/main/java/com/capitalone/dashboard/model/RequestLog.java]: maybe  just make a RequestLog toString() that spits out what u need for logging so when u have to log it then (and only then) will it generate that big string..

Comment (tabladrum) [inline core/src/main/java/com/capitalone/dashboard/model/RequestLog.java]: yeah.

**Pull request merged**